Make get_bin_path() always raise an exception

samdoran opened the pull request:
##### SUMMARY
Change `get_bin_path()` to always raise an exception. This is more in line with how methods in Python behave rather than having a toggle to control whether or not an exception is raised.

`get_bin_path()` in `AnsibleModule` will retain the same interface for backwards compatibility.

##### ISSUE TYPE
<!--- Pick one below and delete the rest -->
- Bugfix Pull Request


##### COMPONENT NAME
<!--- Write the short name of the module, plugin, task or feature below -->
`lib/ansible/module_utils/basic.py`

samdoran committed:
Make get_bin_path always rais an exception

This makes it behave in a more idiomatic way

bcoca reviewed (changes requested):
we use this in several places directly,  you should also change the rest of the callers

samdoran commented:
@bcoca Absolutely. I plan to do that shortly.

samdoran committed:
A few more changes for things using get_bin_path()

samdoran committed:
Use error message from exception when it is redundant

samdoran committed:
Fix bug in Darwin facts for free memory

If the vm_stat command is not found, fact gathering would fail with an unhelpful error message. Handle this gracefully and return a default value for free memory.

samdoran committed:
Change uses of get_bin_path

Catch and handle exception rather than using required param

samdoran committed:
Add changelog

samdoran committed:
Add unit tests

samdoran commented on lib/ansible/module_utils/facts/hardware/darwin.py:
```diff
@@ -85,7 +85,8 @@ def get_cpu_facts(self):
         memory_facts = {
-            'memtotal_mb': int(self.sysctl['hw.memsize']) // 1024 // 1024
+            'memtotal_mb': int(self.sysctl['hw.memsize']) // 1024 // 1024,
+            'memfree_mb': 0,
```
I added this so that we always return a value for `ansible_memfree_mb`. In the current code in `devel`, if `vm_stat` isn't found, that will set `vm_stat_command` to `None`, which gets passed to `run_command()`. That results in the following error:

```
localhost | FAILED! => {
    "changed": false,
    "cmd": null,
    "msg": "Argument 'args' to run_command must be list or string",
    "rc": 257
}
```

So I kept the same behavior, which is setting `vm_stat_command` to `None`, but I added a conditional that avoids calling `run_command()` with a value of `None`.

I could probably do this differently, but that's the reasoning behind this change.

abadger commented on lib/ansible/module_utils/facts/packages.py:
```diff
@@ -79,5 +79,8 @@ def __init__(self):
+            self._cli = get_bin_path(self.CLI)
+        except ValueError:
+            self._cli = None
         return bool(self._cli)
```
This is fine for porting but for greenfield code or when someone next cleans this code up, it would be better to write it more like this:

``` python
except ValueError:
    return False
return True
```

Another possibility is that this whole function would be deemed unneeded in a refactor.  Now that you've fixed up `get_bin_path()` to be more sane the caller of `is_available()` might just want to catch the ValueError itself.

abadger commented on lib/ansible/modules/packaging/os/package_facts.py:
```diff
@@ -255,8 +261,11 @@ def is_available(self):
+                    if get_bin_path(exe):
+                        module.warn('Found "%s" but %s' % (exe, missing_required_lib('apt')))
+                        break
+                except ValueError:
```
I think the ValueError should continue rather than break.  The logic below is what I think is correct (and I also took the opportunity to get rid of the conditional:

``` python
for exe in ('apt', 'apt-get', 'aptitude'):
    try:
        get_bin_path(exe)
        module.warn('Found "%s" but %s' % (exe, missing_required_lib('apt')))
        break
    except ValueError:
        continue
```
I'm not sure if removing the conditional altogether reads clearer than the original.  I might do it this way as a compromise:

``` python
for exe in ('apt', 'apt-get', 'aptitude'):
    try:
        get_bin_path(exe)
    except ValueError:
        continue
    else:
        module.warn('Found "%s" but %s' % (exe, missing_required_lib('apt')))
        break
```

samdoran replied:
Going with the second suggestion because I think it reads better.

abadger commented on lib/ansible/module_utils/facts/hardware/darwin.py:
```diff
@@ -85,38 +85,43 @@ def get_cpu_facts(self):
+            vm_stat_command = get_bin_path('vm_stat')
+        except ValueError:
+            vm_stat_command = None
+        if vm_stat_command:
```
I'm just going to leave this comment once but it applies to most of the ports:

I think this is fine for a port but the next time this is refactored, it should probably look more like this:
``` python
try:
    vm_stat_command = get_bin_path('vm_stat')
except ValueError:
    # The vm stat was not found.  Return early with the values we were able to determine
    return memory_facts

rc, out, err = self.module.run_command(vm_stat_command)
[...]
```

returning early is better code as it allows the reader of the code to stop worrying about what the other conditional path might do later on in the function.  (Learned this from Jack's talk at Pycon 2018)

samdoran replied:
Agree. I like this much better. You can stop reading at the `return` rather than having to read through the rest of the function.

abadger commented on lib/ansible/module_utils/common/process.py:
```diff
@@ -9,13 +9,13 @@
+    Find system executable in PATH. Raises ValueError if executable is not found.
     Optional arguments:
-       - required:  if executable is not found and required is true it produces an Exception
+       - required:  [Deprecated in 2.12] if executable is not found and required is true it produces an Exception
```
The code implementing this has been removed.  So this docstring should be updated to say
```suggestion
       - required:  [Deprecated] Prior to 2.10, if executable is not found and required is true it produces an Exception.  In 2.10 and beyond, we always raise an exception, ignoring this parameter.  This parameter will be removed in 2.12.
```

(Change the version if applicable)

Alternately, the code should be changed to keep honoring the required parameter.  If you do it this way, you might also want to use the deprecated functionality to mark this once you merge your code for doing that without access to an AnsibleModule.

samdoran replied:
That would be good use of deprecations outside of an AnsibleModule object. I'll just update the doc string for now.

samdoran committed:
Update docstring

samdoran committed:
Improve readability

- return early rather than using None with conditionals
- return True or False rather than converting with bool() because it's more straightforward

abadger commented:
Changes look great.  +1 to merge from me!
